# -*- coding: utf-8 -*-
"""
ALPHA SOVEREIGN - CORTEX UPLINK BRIDGE v2.0
=================================================================
Component: shield/nexus/bridge/brain_uplink.py
Core Responsibility: Ø§Ù„Ø¹ØµØ¨ Ø§Ù„Ø¨ØµØ±ÙŠ Ø§Ù„Ø±Ø§Ø¨Ø· Ø¨ÙŠÙ† Ø§Ù„Ø¯Ø±Ø¹ ÙˆØ§Ù„Ø¹Ù‚Ù„.
Forensic Features:
  - Smart Import Discovery (Ø§Ù„Ø¨Ø­Ø« Ø¹Ù† Ø§Ù„Ø¨Ø±ÙˆØªÙˆÙƒÙˆÙ„Ø§Øª ÙÙŠ Ù…Ø³Ø§Ø±Ø§Øª Ù…ØªØ¹Ø¯Ø¯Ø©).
  - Circuit Breaker Pattern (Ø­Ù…Ø§ÙŠØ© Ø§Ù„Ù†Ø¸Ø§Ù… Ù…Ù† Ø§Ù„Ø§Ù†Ù‡ÙŠØ§Ø± Ø§Ù„Ù…ØªØ³Ù„Ø³Ù„).
  - Auto-Healing Connection (Ø¥Ø¹Ø§Ø¯Ø© Ø§Ù„Ø§ØªØµØ§Ù„ Ø§Ù„Ø°Ø§ØªÙŠ).
  - Binary Protocol Logging (ØªØ³Ø¬ÙŠÙ„ Ø®Ø§Ù… Ù„Ù„Ø±Ø³Ø§Ø¦Ù„ Ù„Ù„ØªØ­Ù„ÙŠÙ„ Ø§Ù„Ø¬Ù†Ø§Ø¦ÙŠ).
=================================================================
"""

import logging
import grpc
import asyncio
import sys
import os
import time
from typing import Dict, Any, Optional
from pathlib import Path

# --- 1. Ø§ÙƒØªØ´Ø§Ù Ø§Ù„Ø¨Ø±ÙˆØªÙˆÙƒÙˆÙ„Ø§Øª Ø§Ù„Ø°ÙƒÙŠ (The Fix) ---
# Ù†Ø­Ø§ÙˆÙ„ Ø§Ù„Ø¹Ø«ÙˆØ± Ø¹Ù„Ù‰ Ø§Ù„Ù…Ø¯ÙŠÙˆÙ„Ø§Øª Ø§Ù„Ù…ÙˆÙ„Ø¯Ø© Ø³ÙˆØ§Ø¡ ÙƒØ§Ù†Øª ÙÙŠ brain Ø£Ùˆ shield
generated_paths = [
    Path(__file__).resolve().parent.parent.parent.parent.parent / "brain" / "generated", # brain/generated
    Path(__file__).resolve().parent.parent.parent / "core" / "proto" # shield/core/proto (Backup)
]

brain_pb2 = None
brain_grpc = None

for path in generated_paths:
    if path.exists():
        sys.path.insert(0, str(path))
        try:
            # Ù†Ø­Ø§ÙˆÙ„ Ø§Ø³ØªÙŠØ±Ø§Ø¯ Ø§Ù„Ù†Ø³Ø®Ø© Ø§Ù„Ø£Ø­Ø¯Ø« (EngineControl Ù‡Ùˆ Ø§Ù„Ø§Ø³Ù… Ø§Ù„Ø¬Ø¯ÙŠØ¯)
            import engine_control_pb2
            import engine_control_pb2_grpc
            brain_pb2 = engine_control_pb2
            brain_grpc = engine_control_pb2_grpc
            print(f"âœ… Bridge linked to Neural Protocol at: {path}")
            break
        except ImportError:
            continue

# Ø¥Ø¹Ø¯Ø§Ø¯ Ø§Ù„Ø³Ø¬Ù„Ø§Øª
logger = logging.getLogger("Alpha.Bridge.Uplink")

class CircuitBreaker:
    """Ù‚Ø§Ø·Ø¹ Ø§Ù„Ø¯Ø§Ø¦Ø±Ø© Ù„Ø­Ù…Ø§ÙŠØ© Ø§Ù„Ù†Ø¸Ø§Ù… Ù…Ù† Ø§Ù„Ø·Ù„Ø¨Ø§Øª Ø§Ù„ÙØ§Ø´Ù„Ø© Ø§Ù„Ù…ØªÙƒØ±Ø±Ø©"""
    def __init__(self, failure_threshold=5, recovery_timeout=30):
        self.failures = 0
        self.threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.last_failure_time = 0
        self.state = "CLOSED" # CLOSED=Normal, OPEN=Broken

    def record_failure(self):
        self.failures += 1
        self.last_failure_time = time.time()
        if self.failures >= self.threshold:
            self.state = "OPEN"
            logger.warning("ğŸ”¥ Circuit Breaker OPENED. Pausing Uplink.")

    def record_success(self):
        if self.state == "HALF_OPEN":
            self.state = "CLOSED"
            self.failures = 0
            logger.info("âœ… Circuit Breaker CLOSED. Uplink Restored.")
        elif self.state == "CLOSED":
            self.failures = 0

    def allow_request(self) -> bool:
        if self.state == "CLOSED":
            return True
        if self.state == "OPEN":
            if time.time() - self.last_failure_time > self.recovery_timeout:
                self.state = "HALF_OPEN"
                logger.info("âš ï¸ Circuit Breaker HALF-OPEN. Testing Connection...")
                return True
            return False
        return True # HALF_OPEN allows 1 request

class BrainUplink:
    """
    Ø¬Ø³Ø± Ø§Ù„Ø§ØªØµØ§Ù„ Ø§Ù„Ø¹ØµØ¨ÙŠ.
    """

    def __init__(self):
        # Ù‚Ø±Ø§Ø¡Ø© Ø§Ù„Ø¹Ù†ÙˆØ§Ù† Ù…Ù† Ù…ØªØºÙŠØ±Ø§Øª Ø§Ù„Ø¨ÙŠØ¦Ø© Ø£Ùˆ Ø§Ø³ØªØ®Ø¯Ø§Ù… Ø§Ù„Ø§ÙØªØ±Ø§Ø¶ÙŠ
        self.target = os.getenv("BRAIN_GRPC_TARGET", "localhost:50051")
        self.channel = None
        self.stub = None
        self.breaker = CircuitBreaker()
        self.is_connected = False
        
        if not brain_pb2:
            logger.critical("âŒ NEURAL PROTOCOLS MISSING. Bridge is effectively blind.")

    async def connect(self):
        """Ø¨Ø¯Ø¡ Ø§Ù„Ø§ØªØµØ§Ù„ (ØºÙŠØ± Ù…ØªØ²Ø§Ù…Ù†)"""
        if not brain_pb2: return False
        
        logger.info(f"ğŸ”Œ Initializing Uplink to {self.target}...")
        self.channel = grpc.aio.insecure_channel(self.target)
        self.stub = brain_grpc.EngineControlStub(self.channel)
        
        # Ù…Ø­Ø§ÙˆÙ„Ø© Ø§Ù„Ø§ØªØµØ§Ù„ Ø§Ù„Ø£ÙˆÙ„ÙŠØ©
        if await self._check_health():
            return True
        return False

    async def _check_health(self) -> bool:
        """ÙØ­Øµ Ø§Ù„Ù†Ø¨Ø¶"""
        try:
            await asyncio.wait_for(self.channel.channel_ready(), timeout=2.0)
            self.is_connected = True
            return True
        except asyncio.TimeoutError:
            logger.warning(f"âš ï¸ Uplink Timeout ({self.target}). Brain might be sleeping.")
            return False
        except Exception as e:
            logger.error(f"âŒ Uplink Error: {e}")
            return False

    async def send_signal(self, method: str, payload: Dict) -> Dict:
        """
        Ø¥Ø±Ø³Ø§Ù„ Ø¥Ø´Ø§Ø±Ø© Ø¹Ø§Ù…Ø©.
        ØªÙ‚ÙˆÙ… Ù‡Ø°Ù‡ Ø§Ù„Ø¯Ø§Ù„Ø© Ø¨ØªØºÙ„ÙŠÙ Ø§Ù„Ø·Ù„Ø¨ ÙˆØªÙ…Ø±ÙŠØ±Ù‡ Ø¹Ø¨Ø± Ù‚Ø§Ø·Ø¹ Ø§Ù„Ø¯Ø§Ø¦Ø±Ø©.
        """
        if not self.breaker.allow_request():
            return {"error": "CIRCUIT_OPEN", "msg": "Connection paused due to failures."}

        if not self.is_connected or not self.stub:
            # Ù…Ø­Ø§ÙˆÙ„Ø© Ø¥Ø¹Ø§Ø¯Ø© Ø§Ù„Ø§ØªØµØ§Ù„ Ø§Ù„Ø³Ø±ÙŠØ¹
            if not await self._check_health():
                self.breaker.record_failure()
                return {"error": "DISCONNECTED", "msg": "Brain unreachable."}

        try:
            # 1. ØªÙˆØ¬ÙŠÙ‡ Ø§Ù„Ø·Ù„Ø¨ Ø­Ø³Ø¨ Ø§Ù„Ø¯Ø§Ù„Ø© Ø§Ù„Ù…Ø·Ù„ÙˆØ¨Ø©
            if method == "EXECUTE_ORDER":
                # Ø¨Ù†Ø§Ø¡ Ø§Ù„Ø±Ø³Ø§Ù„Ø© (Mapping Dict -> Protobuf)
                # Ù…Ù„Ø§Ø­Ø¸Ø©: Ø§Ù„Ù‚ÙŠÙ… ÙŠØ¬Ø¨ Ø£Ù† ØªÙƒÙˆÙ† strings ÙÙŠ Ø§Ù„Ø¨Ø±ÙˆØªÙˆÙƒÙˆÙ„ Ø§Ù„Ù…Ø­Ø¯Ø« Ù„Ù„Ø­ÙØ§Ø¸ Ø¹Ù„Ù‰ Ø§Ù„Ø¯Ù‚Ø©
                req = brain_pb2.ExecuteOrderRequest(
                    order_id=payload.get("id", "UNKNOWN"),
                    symbol=payload.get("symbol", ""),
                    side=0 if payload.get("side") == "BUY" else 1,
                    quantity=str(payload.get("qty", "0")),
                    price=str(payload.get("price", "0")),
                    order_type=1 # LIMIT
                )
                
                response = await self.stub.ExecuteOrder(req, timeout=3.0)
                self.breaker.record_success()
                return {"status": "SENT", "server_msg": response.message}
            
            # ÙŠÙ…ÙƒÙ† Ø¥Ø¶Ø§ÙØ© Ø§Ù„Ù…Ø²ÙŠØ¯ Ù…Ù† Ø§Ù„Ø¯ÙˆØ§Ù„ Ù‡Ù†Ø§ (PING, STATUS, etc.)

        except grpc.RpcError as e:
            self.breaker.record_failure()
            logger.error(f"âš¡ RPC Fail: {e.code()} - {e.details()}")
            return {"error": "RPC_FAIL", "code": str(e.code())}
            
        return {"error": "UNKNOWN_METHOD"}

    async def close(self):
        if self.channel:
            await self.channel.close()
            logger.info("ğŸ”Œ Uplink Severed.")

# =================================================================
# Forensic Verification
# =================================================================
if __name__ == "__main__":
    async def test_bridge():
        bridge = BrainUplink()
        
        print(f"[*] Uplink Target: {bridge.target}")
        if not brain_pb2:
            print("[-] Protocols: MISSING (Run proto_compiler.py first)")
            return

        print("[*] Connecting...")
        connected = await bridge.connect()
        
        if connected:
            print("[+] Connection: ESTABLISHED")
            
            # Test Order
            order = {"id": "TEST-01", "symbol": "BTCUSDT", "side": "BUY", "qty": 0.01, "price": 95000}
            print(f"[*] Sending Test Signal: {order}")
            
            res = await bridge.send_signal("EXECUTE_ORDER", order)
            print(f"[+] Result: {res}")
            
        else:
            print("[-] Connection: FAILED (Is the Brain running?)")
            
        await bridge.close()

    try:
        asyncio.run(test_bridge())
    except KeyboardInterrupt:
        pass